### DISCUSSION TITLE
How to clearly separate unit tests and integration tests?

### DISCUSSION THREAD

**Discussion by lppedd** [eclipse-tycho/tycho#1716] [Q&A] [answered]
Currently the testing strategies offered by Tycho seems to be a bit of mix up. It's difficult to get a sense out of it.

- `tycho-surefire-plugin:test` is executed only for `eclipse-test-plugin` packaging, in the `integration-test` phase.  
   This is strange for two reasons: 
   1. I would expect a `test` mojo to run on the `test` phase
   2. It only applies to `eclipse-test-plugin`, and not to `eclipse-plugin`

- `tycho-surefire-plugin:plugin-test` is executed only for `eclipse-plugin` packaging, in the `integration-test` phase.  
   Feel strange because I'd like to execute JUnit tests on the `test` phase, and only SWTBot tests on the `integration-test` phase.

Ultimately my question is, which approach is the modern approach? How do I end up with unit tests in the same source bundle executed in the `test` phase, and SWTBot tests in the `integration-test` phase?

**Comment by laeubi** [comment, marked as answer]
> Currently the testing strategies offered by Tycho seems to be a bit of mix up. It's difficult to get a sense out of it.

Yep we track this here:
- #1502 
- https://github.com/eclipse-tycho/tycho/discussions/1597
 
>     * `tycho-surefire-plugin:test` is executed only for `eclipse-test-plugin` packaging, in the `integration-test` phase.
>       This is strange for two reasons:
>       
>       1. I would expect a `test` mojo to run on the `test` phase
>       2. It only applies to `eclipse-test-plugin`, and not to `eclipse-plugin`

Because of this I do not  recommend using  `eclipse-test-plugin` packaging for new designs but we have to keep it that way for historic reasons.

>     * `tycho-surefire-plugin:plugin-test` is executed only for `eclipse-plugin` packaging, in the `integration-test` phase.
>       Feel strange because I'd like to execute JUnit tests on the `test` phase, and only SWTBot tests on the `integration-test` phase.

The default binding was removed in Tycho 4.x because it is not flexible enough, even though you can bind "classic" JUnit Tests...

> Ultimately my question is, which approach is the modern approach? How do I end up with unit tests in the same source bundle executed in the `test` phase, and SWTBot tests in the `integration-test` phase?

... see for example the following integration-tests

https://github.com/eclipse-tycho/tycho/tree/master/tycho-its/projects/surefire.combinedtests

**Reply by lppedd** [reply]
Thank you! Looking at the IT POM helped.
So, to recap:
1. include unit/integration tests in `eclipse-plugin` packaging
2. use `maven-surefire-plugin` bind to the `test` phase (which is by default) to execute unit tests without a workbench
    https://github.com/eclipse-tycho/tycho/blob/42397c7d0514e51fff3d39e2a66c98375123b8f6/tycho-maven-plugin/src/main/resources/META-INF/plexus/components.xml#L39-L41
3. use `tycho-surefire-plugin` bind to `integration-test` to execute SWTBot tests.

However, I then don't get why we have `<useUIHarness>` in `tycho-surefire-plugin` if we expect to run only integration tests. I would expect it to be true by default.

**Reply by laeubi** [reply]
> However, I then don't get why we have <useUIHarness> in tycho-surefire-plugin if we expect to run only integration tests. I would expect it to be true by default.

Integration test run inside an OSGi framework, they do not necessarily have an UI and even if they have an UI the do not necessarily require `useUIHarness` (what is a special flag for Eclipse Ui tests!).

**Reply by lppedd** [reply]
@laeubi got it. This discussion helped too https://github.com/eclipse-tycho/tycho/discussions/1492#discussioncomment-3832137
Seems like I was not the only one confused at least 😆
